apps/virtual_joystick/main.py

```diff
@@ -0,0 +1,410 @@
+        self.amiga_rate = str(self.amiga_tpdo1.meas_ang_rate)
+
+    def on_exit_btn(self) -> None:
+        App.get_running_app().stop()
+
+    async def draw_joystick(self) -> None:
+        """Loop over drawing the VirtualJoystickWidget."""
+        while self.root is None:
+            await asyncio.sleep(0.01)
+        joystick: VirtualJoystickWidget = self.root.ids["joystick"]
+        while True:
+            joystick.draw()
+            await asyncio.sleep(0.01)
+
+    async def app_func(self):
+        async def run_wrapper() -> None:
+            # we don't actually need to set asyncio as the lib because it is
+            # the default, but it doesn't hurt to be explicit
+            await self.async_run(async_lib="asyncio")
+            for task in self.async_tasks:
+                task.cancel()
+
+        # configure the canbus client
+        canbus_config: CanbusClientConfig = CanbusClientConfig(
+            address=self.address, port=self.canbus_port
+        )
+        canbus_client: CanbusClient = CanbusClient(canbus_config)
+
+        # configure the camera client
+        camera_config: OakCameraClientConfig = OakCameraClientConfig(
+            address=self.address, port=self.camera_port
+        )
+        camera_client: OakCameraClient = OakCameraClient(camera_config)
+
+        # Drawing task(s)
+        self.async_tasks.append(asyncio.ensure_future(self.draw_joystick()))
+
+        # Canbus task(s)
+        self.async_tasks.append(
+            asyncio.ensure_future(self.stream_canbus(canbus_client))
+        )
+        self.async_tasks.append(
+            asyncio.ensure_future(self.send_can_msgs(canbus_client))
+        )
+        self.async_tasks.append(
+            asyncio.ensure_future(canbus_client.poll_service_state())
+        )
+
+        # Camera task(s)
+        self.async_tasks.append(
+            asyncio.ensure_future(self.stream_camera(camera_client))
+        )
+        self.async_tasks.append(
+            asyncio.ensure_future(camera_client.poll_service_state())
+        )
+
+        return await asyncio.gather(run_wrapper(), *self.async_tasks)
+
+    async def stream_camera(self, client: OakCameraClient) -> None:
+        """This task listens to the camera client's stream and populates the tabbed panel with all 4 image streams
+        from the oak camera."""
+        while self.root is None:
+            await asyncio.sleep(0.01)
+
+        response_stream: Optional[Generator[oak_pb2.StreamFramesReply]] = None
+
+        while True:
+            while client.state.value != oak_pb2.OakServiceState.RUNNING:
+                # start the streaming service
+                await client.start_service()
+                await asyncio.sleep(0.01)
+
+            if response_stream is None:
+                # get the streaming object
+                response_stream = client.stream_frames(every_n=self.stream_every_n)
+                await asyncio.sleep(0.01)
+                continue
```

Comment:
@edgarriba thoughts on the extra sleep and continue? If unnecessary I will remove
```suggestion
            while client.state.value != oak_pb2.OakServiceState.RUNNING:
                # start the streaming service
                await client.start_service()

            if response_stream is None:
                # get the streaming object
                response_stream = client.stream_frames(every_n=self.stream_every_n)
```

Reply:
i would test running the app and see performance to be sure